### DISCUSSION TITLE
A word on Analytics

### DISCUSSION THREAD

**Discussion by 444B** [444B/streamlit-analytics2#120] [Performance Engineering]
I have been trying to integrate other analytics tools into SA2 for a while now and have tried every method conceivable and described in the forums.
I have not succedded to get a little bit of JS to work.

However, In my recent attempts to get it functional - I was looking at the `st.context` function and found some very interesting things.

```python
import streamlit as st

st.title("HTTP Headers")
st.context.headers

st.title("HTTP Cookies")
st.context.cookies
```

The following is provided from the client to your streamlit server with anytime someone views your page or presses a button   

HTTP Headers
```json
{
   "Host":"REDACTED",
   "User-Agent":"REDACTED",
   "Accept-Encoding":"REDACTED",
   "Accept-Language":"REDACTED",
   "Cache-Control":"REDACTED",
   "Cookie":"ajs_anonymous_id=REDACTED; _streamlit_xsrf=REDACTED; _streamlit_xsrf=REDACTED",
   "Origin":"REDACTED",
   "Pragma":"no-cache",
   "X-Request-Id":"REDACTED",
   "X-Real-Ip":"REDACTED",
   "X-Forwarded-Port":"REDACTED",
   "X-Forwarded-Scheme":"REDACTED",
   "X-Original-Uri":"REDACTED",
   "X-Scheme":"REDACTED",
   "Sec-Websocket-Version":"REDACTED",
   "Sec-Websocket-Key":"REDACTED",
   "Sec-Websocket-Extensions":"REDACTED",
   "Sec-Websocket-Protocol":"REDACTED",
   "X-Original-Proto":"REDACTED",
   "X-Forwarded-Proto":"REDACTED",
   "X-Forwarded-Host":"REDACTED",
   "X-Forwarded-For":"REDACTED",
   "Proxy-Connection":"REDACTED",
   "Connection":"REDACTED",
   "Upgrade":"REDACTED"
}
```

Cookies
```json
{
   "ajs_anonymous_id":"REDACTED",
   "_streamlit_xsrf":"REDACTED"
}
```

I beleive the way to prevent these cookies and the ajs_anonymous_id from being embedded in your client is to set  
`[browser] gatherUsageStats = false`   
in the `.steamlit/config.toml`. I have tested that and it works.

I do wonder if we could use these values to make SA2 more meaningful however a key point here is that analytics seems to be easy for Streamlit to do for themselves but impossible for any users of streamlit.

**Comment by emigre459** [comment]
Kind of related, but I've been trying to get analytics using streamlit-analytics at the session/user level instead of what seems to be in aggregate across all sessions/users (I'm still testing locally, so please correct me if that's wrong, but I don't see any unique ID that would track clicks at the session level). Is there any planned functionality for that? I'm currently tracking visitors by using UUID session IDs and setting a cookie in the client browser and then using firestore to track user-level clicks, but it's not playing nicely with other packages currently (namely `streamlit-authenticator` or perhaps the underlying CookieManager).

Would there be any interest in a PR that could for example take a `visitor_id` string as a kwarg to `track` and `start_tracking` that could then be used as a document ID in place of `"counts"` in Firestore? I'm not sure how it would need to adjust (or even if it could) to also work for the non-firestore use case though...
